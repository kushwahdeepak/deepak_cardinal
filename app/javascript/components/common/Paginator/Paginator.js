import React, { useMemo, useRef } from 'react'
import Pagination from 'react-bootstrap/Pagination'

import './styles/Paginator.scss'

function Paginator(props) {
    const inputRef = useRef()
    const {
        pageCount,
        pageWindowSize,
        activePage,
        setActivePage,
        showGoToField = true,
    } = props
    const lastPageIndex = pageCount - 1

    const handleChange = (evt) => {
        const notAllowedKeysCode = [46, 43, 69, 101, 45]
        const code = evt.which || evt.keyCode

        if (notAllowedKeysCode.includes(code)) {
            evt.preventDefault()
        }

        if (evt.key === 'Enter') {
            try {
                goToPage(evt.target.value)
            } catch (error) {
                console.log(error)
            }
        }
    }

    const goToPage = (page) => {
        const parsedNumber = Number(page) - 1
        if (parsedNumber < 0 || parsedNumber > lastPageIndex) return
        setActivePage(parsedNumber)
        if (inputRef.current) inputRef.current.value = ''
    }

    const visiblePaginationItems = useMemo(() => {
        const { rangeStart, rangeEnd } = calcVisibleItemRange(
            pageWindowSize,
            pageCount,
            activePage
        )

        const prevPageIndex = Math.max(activePage - 1, 0)
        const nextPageIndex = Math.min(activePage + 1, lastPageIndex)

        const items = []

        items.push(
            <Pagination.First
                key="prev5"
                className="prev"
                onClick={() =>
                    setActivePage(Math.max(activePage - pageWindowSize, 0))
                }
            />
        )

        items.push(
            <Pagination.Prev
                key="prev"
                className="prev"
                onClick={() => setActivePage(prevPageIndex)}
            />
        )

        items.push(
            <Pagination.Item
                key="first"
                data-test="first-page-btn"
                active={activePage === 0}
                onClick={() => setActivePage(0)}
            >
                1
            </Pagination.Item>
        )

        if (rangeStart >= 2)
            items.push(<Pagination.Ellipsis key={'start-ellipsis'} />)

        for (let i = rangeStart; i <= rangeEnd; i++) {
            if (i === lastPageIndex || i === 0) continue
            items.push(
                <Pagination.Item
                    key={i}
                    active={i === activePage}
                    onClick={() => setActivePage(i)}
                >
                    {i + 1}
                </Pagination.Item>
            )
        }

        if (rangeEnd + 1 < pageCount - 1)
            items.push(<Pagination.Ellipsis key={'end-ellipsis'} />)

        items.push(
            <Pagination.Item
                key="last"
                data-test="last-page-btn"
                active={lastPageIndex === activePage}
                onClick={(event) => setActivePage(lastPageIndex)}
            >
                {lastPageIndex + 1}
            </Pagination.Item>
        )

        items.push(
            <input
                key="to_page"
                type="number"
                min="0"
                placeholder="To page: "
                ref={inputRef}
                onKeyPress={(evt) => handleChange(evt)}
                className="paginator-input"
            />
        )

        items.push(
            <Pagination.Next
                key="next"
                onClick={() => setActivePage(nextPageIndex)}
            />
        )

        items.push(
            <Pagination.Last
                key="next5"
                onClick={() =>
                    setActivePage(
                        Math.min(activePage + pageWindowSize, lastPageIndex)
                    )
                }
            />
        )

        return items
    }, [pageCount, activePage])

    return pageCount && pageCount > 1 ? (
        <Pagination data-test="pagination">{visiblePaginationItems}</Pagination>
    ) : null
}

function calcVisibleItemRange(windowSize, pageCount, activePage) {
    const lastPageIndex = pageCount - 1
    const rangeWidth = Math.min(windowSize, pageCount)
    let rangeStart = Math.max(activePage - Math.floor(rangeWidth / 2), 0)
    const rangeEnd = Math.min(rangeStart + rangeWidth - 1, lastPageIndex)
    const rangeEndOverextension = Math.max(
        rangeStart + rangeWidth - 1 - lastPageIndex,
        0
    )
    rangeStart -= rangeEndOverextension

    return { rangeStart, rangeEnd }
}

export default Paginator
