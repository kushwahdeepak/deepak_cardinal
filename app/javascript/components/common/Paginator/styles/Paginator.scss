.page-item a span {
    color: #5263be;
}

.page-item a {
    background-color: transparent;
    border: none !important;
    color: #5263be !important;
    font-weight: bold;
    margin: 0 10px;
}

.page-item span {
    border: none !important;
    margin: 0 10px;
}

.pagination .active .page-link {
    background: linear-gradient(94.67deg, #4C68FF -1.19%, #6077F4 53.94%, #8185FF 102.59%);
    border-radius: 5px;
    color: white;
}

.paginator-input {
    width: 70px;
    color: #5263be;
    font-weight: bold;
    border-radius: 10px;
    border-color:  #5263be;
    &::placeholder {
        font-size: 10px;
        font-weight: bold;
        color: #5263be;
    }
    text-align: center;
    padding-bottom: 5px;
}

@media (max-width: 500px) {
    .page-item a {
        margin: 0;
    }
}

@media only screen and (min-device-width : 320px) and (max-device-width : 575px){
    .paginator-input{
        width: 50px;
    }
    .page-item a {
        padding: 0.5rem 5px !important;
    }
}