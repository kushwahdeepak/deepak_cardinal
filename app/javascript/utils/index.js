import { act } from 'react-dom/test-utils'

export const findByTestAttr = (wrapper, attr) => {
    return wrapper.find(`[data-test='${attr}']`)
}

export const timeoutResolve = (time = 0, response = {}) =>
    new Promise((resolve) => {
        setTimeout(() => resolve(response), time)
    })

export const wait = (time = 0) => {
    const timer = timeoutResolve(time)

    return act(async () => {
        await timer
    })
}
export function emailIsValid(email) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)
}

export function generateRandom(min, max) {
    return Math.floor(min + Math.random() * (max + 1 - min))
}

export function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase()
}

export function createMarkedList(array) {
    if (!Array.isArray(array)) {
        return []
    }
    return array.map((candidate) => ({
        id: candidate.id,
        name: candidate.id,
        fullName: `${candidate.first_name} ${candidate.last_name}`,
        emailAddress: candidate.email_address,
        isChecked: false,
    }))
}

export function stringReplace(string,toSymbol,fromSymbol){
    return string.replace(toSymbol, fromSymbol)
}

export function capitalize(string) {
    return string.replace(/\b(\w)/g, s => s.toUpperCase());
}

export function skillIsNotEmpty(value){
    return value.trim().length > 0;
}

export function firstCharacterCapital(value){
    return value?.charAt(0)?.toUpperCase() + value?.slice(1)
}
