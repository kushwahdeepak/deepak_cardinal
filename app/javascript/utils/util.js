import isNil from 'lodash.isnil'
import isEmpty from 'lodash.isempty'

class Util {
    /**
     * Define stage values
     */
    static STAGES = ['first_interview', 'second_interview', 'offer', 'reject']

    /**
     * Tests whether a single object attribute includes the at
     * least one string from the given set.
     */
    static attributePassesStringSetFilter(object, attr, stringSet) {
        if (isNil(object[attr])) return false

        // We "OR" the tag tests: the attribute needs to contain only one tag
        return stringSet
            .map((string) => string.trim().toLowerCase())
            .reduce(
                (passes, string) =>
                    passes ||
                    Util.attributeToString(object[attr])
                        .toLowerCase()
                        .includes(string),
                false
            )
    }

    /**
     * Tests whether at least one of a set of object attributes
     * includes every word from a given piece of text.
     */
    static objectPassesTextFilter(object, attrs, text) {
        text = text.trim().toLowerCase()

        if (text === '') return true

        const words = text.split(' ')

        // Join all the given attribute names to lowercase, stringified attribute values
        const allAttrsString = attrs
            .filter((attr) => !isNil(object[attr]))
            .map((attr) => Util.attributeToString(object[attr]).toLowerCase())
            .join('')

        // Is every word included in the allAttrsString?
        return words.reduce(
            (matchesAll, word) => matchesAll && allAttrsString.includes(word),
            true
        )
    }

    static attributeToString(attr) {
        if (typeof attr === 'string') {
            return attr
        } else if (Array.isArray(attr)) {
            return attr.join(',')
        } else if (typeof attr === 'object') {
            return JSON.stringify(attr)
        } else {
            return '' + attr
        }
    }

    static numberWithCommas(x) {
        return x?.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
    }

    /**
     * Returns number of results in format [x] - [y] of [z]
     * @param {number} totalJobsCount Total number of jobs
     * @param {number} pageCount Total number of pages
     * @param {number} activePage Currently active page
     * @param {number} pageSize How many results to display per page
     */
    static displayNumberOfResults(
        totalJobsCount,
        pageCount,
        activePage,
        pageSize,
        totalPersonCount=null,
        candidatePage=false,
        showdisplaycount=true
    ) {
        if (totalJobsCount == 0) return '0 - 0 of 0'
        const offset = pageSize * pageCount - totalJobsCount
        const lastJobIndex =
            (activePage + 1) * pageSize > totalJobsCount
                ? totalJobsCount
                : (activePage + 1) * pageSize
        const firstJobIndex =
            activePage == 0
                ? 1
                : activePage == pageCount - 1
                ? lastJobIndex - pageSize + offset + 1
                : lastJobIndex - pageSize + 1
        return `${firstJobIndex} - ${lastJobIndex} of ${Util.numberWithCommas(
            (candidatePage && showdisplaycount) ? totalPersonCount : totalJobsCount

        )}`
    }

    /**
     * Returns a string indicating the candidate's title and current
     * company, based on available information on the Person object
     * @param {object} candidate Person object from our db
     */
    static candidateCompanyString(candidate) {
        const position = candidate.company_position
        const currentCompany = candidate.current_company

        if (isEmpty(position) && isEmpty(currentCompany)) return ''

        if (!isEmpty(position) && !isEmpty(currentCompany))
            return `${position} at ${currentCompany}`

        if (!isEmpty(position)) return position

        if (!isEmpty(currentCompany)) return `at ${currentCompany}`
    }

    /**
     * Converts the given array of objects with ids to an object whose
     * keys are the ids of the array objects.
     *
     * Every object in the given array must have a unique 'id' property defined.
     *
     * @param {Array} array
     * @returns {Object}
     */
    static arrayToObject(array) {
        const obj = {}
        array.forEach((element) => (obj[element.id] = element))
        return obj
    }

    /**
     * Makes a copy of the given array and replace the value at 'index' with 'value'
     *
     * @param {Array} array
     * @param {Number} index
     * @param {*} value
     * @returns {Array}
     */
    static replaceInArrayCopy_i(array, index, value) {
        const clone = array.slice()
        clone[index] = value
        return clone
    }

    /**
     * Makes a copy of the given array and replaces the element with the same
     * .id property as obj.id with obj
     *
     * @param {Array} array
     * @param {Object} obj
     * @returns {Array}
     */
    static replaceInArrayCopy(array, obj) {
        const index = array.findIndex((element) => element.id === obj.id)
        return Util.replaceInArrayCopy_i(array, index, obj)
    }

    static deleteInArrayCopy(array, index) {
        const copy = array.slice()
        copy.splice(index, 1)

        return copy
    }

    static handleUndefined(value){
        if (value === undefined || value === 'undefined')
        return '';
        else
        return value
    }

    static handleUndefinedFullName(firstName, lastName){
      if(firstName === "undefined" || firstName === undefined)
      return lastName
      else if(lastName === "undefined" || lastName === undefined)
      return firstName
      else
      return firstName + ' ' + lastName     }

    static capitalize(value){
         return value.charAt(0).toUpperCase() + value.slice(1).toLowerCase();
    }
}

export default Util
