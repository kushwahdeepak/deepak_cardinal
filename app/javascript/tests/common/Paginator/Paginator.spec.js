import React from 'react';
import Paginator from '../../../components/common/Paginator/Paginator';
import renderer from 'react-test-renderer';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { findByTestAttr } from '../../../utils';

configure({adapter: new Adapter()});

const initialShallowRender = (props = {}) => {
	return shallow(<Paginator {...props}/>);
};

const initialMountRender = (props = {}) => {
	return mount(<Paginator {...props}/>);
};

const calcVisibleItemRange = jest.fn();

describe('Paginator component',  () => {
	let wrapper;
	let fullWrapper;

	beforeEach(() => {
		const mockCallBack = jest.fn();

		const props = {
			pageCount: 10, pageWindowSize: 5, activePage: 1, setActivePage: mockCallBack
		};
		wrapper = initialShallowRender(props);
		fullWrapper = initialMountRender(props);
	});

	afterEach(() => {
		fullWrapper.unmount();
	});

	it('should render Paginator component', () => {
		expect(wrapper.exists()).toBe(true);
	});
	it('should match with a snapshot', () => {
		const component = renderer.create(<Paginator />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
	it('should contain the root Pagination tag', () => {
		expect(wrapper.prop('data-test')).toBe('pagination');
	});
	it.skip('should get correct rangeStart and rangeEnd', () => {
		const { rangeStart, rangeEnd } = calcVisibleItemRange();
		expect(rangeStart).toBe(0);
		expect(rangeEnd).toBe(4);
	});
	it('should contain the correct number of children', () => {
		expect(wrapper.children().length).toBe(8);
		wrapper.setProps({ pageWindowSize: 10, pageCount: 20 });
		expect(wrapper.children().length).toBe(13);
	});
	it('should call LastPage button', () => {
		const lastPageBtn = findByTestAttr(fullWrapper, 'last-page-btn').at(0);
		lastPageBtn.find('a').simulate('click');
		expect(fullWrapper.props().setActivePage).toHaveBeenCalledTimes(1);
	});
	it('should call FirstPage button', () => {
		const firstPageBtn = findByTestAttr(fullWrapper, 'first-page-btn').at(0);
		firstPageBtn.find('a').simulate('click');
		expect(fullWrapper.props().setActivePage).toHaveBeenCalledTimes(1);
	});
	it('should have correct prop values', () => {
		expect(fullWrapper.props().pageCount).toEqual(10);
		expect(fullWrapper.props().pageWindowSize).toEqual(5);
		expect(fullWrapper.props().activePage).toEqual(1);
	});
	it('should call FirstPage button with correct arguments', () => {
		const firstPageBtn = findByTestAttr(fullWrapper, 'first-page-btn').first();
		const firstPageIndex = 0;
		firstPageBtn.find('a').simulate('click');
		expect(fullWrapper.props().setActivePage).toHaveBeenCalledWith(firstPageIndex);
	});

	it('should call LastPage button with correct arguments', () => {
		const lastPageBtn = findByTestAttr(fullWrapper, 'last-page-btn').first();
		const lastPageIndex = fullWrapper.props().pageCount - 1;
		lastPageBtn.find('a').simulate('click');
		expect(fullWrapper.props().setActivePage).toHaveBeenCalledWith(lastPageIndex);
		// wrapper.update();
		// expect(lastPageBtn.hasClass('active')).toBe(true);
		//
		// expect(wrapper.props().activePage).toEqual(10);
	});
	describe('Paginator items', () => {
		let wrapper;
		let fullWrapper;

		beforeEach(() => {
			const mockCallBack = jest.fn();

			const props = {
				pageCount: 10, pageWindowSize: 5, activePage: 1, setActivePage: mockCallBack
			};
			wrapper = initialShallowRender(props);
			fullWrapper = initialMountRender(props);
		});

		afterEach(() => {
			fullWrapper.unmount();
		});

		it('should have active class after click', () => {
			const lastPageIndex = fullWrapper.props().pageCount - 1;
			fullWrapper.setProps({ activePage: lastPageIndex });
			fullWrapper.update();
			expect(findByTestAttr(findByTestAttr(fullWrapper, 'pagination').first(), 'last-page-btn').first().childAt(0).hasClass('active')).toBe(true);
		});
	});

});
