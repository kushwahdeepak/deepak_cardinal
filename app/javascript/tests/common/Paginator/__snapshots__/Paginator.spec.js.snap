// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator component should match with a snapshot 1`] = `
<ul
  className="pagination"
  data-test="pagination"
>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      >
        ‹
      </span>
      <span
        className="sr-only"
      >
        Previous
      </span>
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      data-test="first-page-btn"
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
    >
      1
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      data-test="last-page-btn"
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
    >
      NaN
    </a>
  </li>
  <li
    className="page-item"
  >
    <a
      className="page-link"
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
    >
      <span
        aria-hidden="true"
      >
        ›
      </span>
      <span
        className="sr-only"
      >
        Next
      </span>
    </a>
  </li>
</ul>
`;
